xarray/core/computation.py

```diff
@@ -1527,6 +1527,132 @@ def dot(*arrays, dims=None, **kwargs):
     return result.transpose(*[d for d in all_dims if d in result.dims])
 
 
+def cross(a, b, spatial_dim=None):
+    """
+    Return the cross product of two (arrays of) vectors.
+
+    Parameters
+    ----------
+    a : array_like
+        Components of the first vector(s).
+    b : array_like
+        Components of the second vector(s).
+    spatial_dim : something
```

Comment:
My intuition for the cross product is lost once we go beyond cartesian dims. But what does it even mean if we do `np.cross(a(cartesian), b(something_else))`?  Or shall I just accept it as math-magics with arrays that have the correct length? I'm used to that too...

Reply:
I meant that the 2 or 3 element dimension might be named differently for `a` and `b`:
```python
a = xr.DataArray([[0, 1, 2], [3, 4, 5], [7, 5, 3]], dims=("dim_0", "dim_1"))
b = xr.DataArray([[1, 4, 2], [8, 1, 3], [5, 4, 9]], dims=("x", "y"))
xr.cross(a, b, dim=("dim_1", "y"))
```

Reply:
But are `dim_1` and `y` actually different things or is it a typo made by the user? 

I just want to argue that since we have named dims and do all the necessary reshaping behind the scenes we don't need these extra options. 
But I suppose I can think of it like this: `dim_1` is a alias of `y`, so `dim=(dim_1, y)` is basically just adding aliases of the same thing.

Reply:
Another thing to think about if we want to accept different dims, what to do when they have different values?
```python
a = xr.DataArray(
    [[0, 1, 2], [3, 4, 5]],
    dims=("dim_0", "dim_1"),
    coords=dict(dim_0=(["dim_0", [1, 2]]), dim_1=(["dim_1"], ["u", "v", "w"])),
)
b = xr.DataArray(
    [[1, 4, 2], [8, 1, 3]],
    dims=("time", "cartesian"),
    coords=dict(time=(["time", [0, 1]]), cartesian=(["cartesian"], ["x", "y", "z"])),
)
xr.cross(a, b, dim=("dim_1", "y"))
```

Reply:
right. I didn't really think this through before suggesting it so we might have to reject the idea.

Another thing: usually, there's not a dimension like `cartesian` but rather a `Dataset` with a variable for each of `x`, `y`, and `z`. Do we want to require stacking those before passing them to `xr.cross` (using `Dataset.to_stacked_array`), or do we want to add a syntax to do that automatically?

Reply:
I'm leaning towards requiring to have them stacked. Mostly because I don't fully understand how to implement it. 
Maybe that would be a fit for `ds.cross()`?

Reply:
as far as I can tell, this would work with:
```python
In [2]: xr.set_options(keep_attrs=True)
   ...: 
   ...: ds = xr.tutorial.open_dataset("eraint_uvz")
   ...: 
   ...: arr = ds.to_stacked_array(variable_dim="cartesian", new_dim="variable", sample_dims=ds.dims).unstack("variable")
   ...: other_ds = arr.stack(variable=["cartesian"]).to_unstacked_dataset("variable")
   ...: xr.testing.assert_equal(ds, other_ds)
```
> Maybe that would be a fit for `ds.cross()`?

Maybe, but usually top-level functions accept both `DataArray` and `Dataset` objects (not sure if this can / should be an exception?).

Reply:
Was it something like in 14decb3 you imagined, @keewis?

Reply:
@keewis, found a dataset example here that I'm not sure how to handle, any ideas?

```python
ds = xr.Dataset({0: ("dim_0", [1]), 1: ("dim_0", [2]), 2: ("dim_0", [3])})

arr = ds.to_stacked_array(variable_dim="cartesian", new_dim="variable", sample_dims=ds.dims).unstack("variable")
other_ds = arr.stack(variable=["cartesian"]).to_unstacked_dataset("variable")
xr.testing.assert_equal(ds, other_ds)

AssertionError: Left and right Dataset objects are not equal
Differing dimensions:
    (dim_0: 1) != ()

Differing data variables:
L   0        (dim_0) int32 1
R   0        int32 1
L   1        (dim_0) int32 2
R   1        int32 2
L   2        (dim_0) int32 3
R   2        int32 3
```

Reply:
sorry for the late reply.

> Was it something like in 14decb3 you imagined, @keewis?

yes, exactly. I'm somewhat uncomfortable with the big loop over the input objects, but I'm not sure what I would suggest instead.

> found a dataset example here that I'm not sure how to handle

yes, 1-sized dimensions can be tricky because they tend to lose their status as a dimension. We can try to use `expand_dims` to get back the original dims:
```python
other_ds.expand_dims([dim for dim, size in ds.sizes.items() if size == 1])
```

Reply:
Here's another case I've been pondering about, should different dims be allowed?
```python
a = xr.Dataset({0: ("dim_0", [1]), 1: ("dim_0", [2]), 2: ("dim_0", [3])})
b = xr.Dataset({0: ("dim_1", [4]), 1: ("dim_1", [5]), 2: ("dim_1", [6])})
c = xr.cross(a, b, "temp_dim")
```

How should c look like? Should it behave like a*b?
```python
a*b
Out[25]: 
<xarray.Dataset>
Dimensions:  (dim_0: 1, dim_1: 1)
Dimensions without coordinates: dim_0, dim_1
Data variables:
    0        (dim_0, dim_1) int32 4
    1        (dim_0, dim_1) int32 10
    2        (dim_0, dim_1) int32 18
```
In bce2f3e it follows the a*b style which is also consistent when size>1.